packages/grid/_modules_/grid/hooks/features/useComponents.tsx

```diff
@@ -31,35 +36,23 @@ export const useComponents = (
     [pagination, rows, columns, options, apiRef, gridRootRef],
   );
 
-  const headerComponent = React.useMemo(
-    () =>
-      componentOverrides?.header
-        ? React.createElement(componentOverrides.header, componentProps)
-        : null,
-    [componentOverrides, componentProps],
-  );
-
-  const footerComponent = React.useMemo(
-    () =>
-      componentOverrides?.footer
-        ? React.createElement(componentOverrides.footer, componentProps)
-        : null,
-    [componentOverrides, componentProps],
-  );
+  const headerComponent = componentOverrides?.Header
+    ? React.createElement(componentOverrides.Header, componentProps)
+    : options.showToolbar && <DefaultToolbar />;
 
   const loadingComponent = React.useMemo(
     () =>
-      componentOverrides?.loadingOverlay ? (
-        React.createElement(componentOverrides.loadingOverlay, componentProps)
+      componentOverrides?.LoadingOverlay ? (
+        React.createElement(componentOverrides.LoadingOverlay, componentProps)
```

Comment:
Why not doing?

```suggestion
        <componentOverrides.LoadingOverlay />
```

Reply:
Or even better, we can remove the branching with a default value for each component.